1st draft at HowTo (#239)
oscar-system/Oscar.jl: HowTo.md

> # Oscar - how do we envision development

Reviewer: IMHO the file should be renamed, perhaps to "PLAN.md" or "DEVELOPMENT.md" or so

## What Do We Have:

> We have a large code-base for infrastructure in place, comprising at least

Reviewer: ```suggestion
We have a large codebase for infrastructure in place, comprising at least
```

 - matrices
 - polynomials (univariate and multivariate)
 - power series
 - number fields
 - (abelian) groups
 - ... and MUCH more
As well as specialised stuff
 - polymake
 - Singular
 - Gap

> So: please use it. It is safe to assume all can be improved, however,
> then we won't get anywhere. For preference: interoperability is much more

Reviewer: ```suggestion
So: please use it. It is safe to assume all can be improved, however, if we try to perfect every
single line of code again and again, we we won't get anywhere; there is a balance to be found.
For preference: interoperability is much more
```

important than speed, readability is much more important than speed. Having
said that: of course, sometimes pure speed matters, but not nearly as often as
people think.

The infrastructure is incomplete, e.g. we do not have

>  - varieties, (we're started with curves recently)

Reviewer: ```suggestion
 - varieties, (we've started with curves recently)
```

 - polytopes, cones
 - surfaces
 - ....

If you need them, e.g. you want a library/ package for (elliptic) curves, then
step ONE should be to define a curve. It does not need to be perfect, support
all of Hartshorn immediately, but it also should not contradict him either. It
is OK for a function to say: sorry, not (yet) implemented. Then if I need it,
I can add and re-use the infrastructure.

The key idea is to pick a single textbook and follow the conventions in there. This
way we get a consistent interface.

## What Is Missing?

At this point we have, mathematically speaking, almost nothing. This has to
change, quickly.  We work in a team of ~ 10(?) people.
We lack infrastructure as everyone will notice immediately.
As a result, the classical
   I work for 6 month in a separate repo on a branch and then will dazzle you with perfect
   code and cool examples
is not going to work for now. It will result in everyone fixing the same
infrastructure problems over and over again. Please consider to work, e.g. in

> a file/directory  in Oscar/examples and push on a regular basis, even, or in

Reviewer: ```suggestion
a file/directory  in `Oscar/examples` and push on a regular basis, even, or in
```

particular, incomplete code. Break it down into small pull requests.
 - it is in examples, so no-one is accidentally going to automatically use it
 - it is in the main repo, so everyone can see it and try to use it, producing feedback
 - infrastructure, in Oscar at least, can be added in the same repo, immediately and thus
   is available to others, immediately. (feedback, ...)
 - Talk to us - as early as possible. There is a chance that what you're trying to
   do is already there - in a different package maybe, not linked in, ....

>  - Talk to us! We are

Reviewer: ```suggestion
 - Talk to us!
```

>    - on slack: send an email to webmaster-oscar@mathematik.uni-kl.de

Reviewer: ```suggestion
   - on slack: send an email to webmaster-oscar@mathematik.uni-kl.de to be added
```

   - on oscar-dev@mathematik.uni-kl.de

>    

Reviewer: ```suggestion
   - whenever you meet us (at the uni, during conferences, ...)
```

# Practical Development

For practical development:

>  - use ```Revise```

Reviewer: ```suggestion
 - use [the `Revise` package](https://github.com/timholy/Revise.jl)
```

 - start a file/ directory in Oscar/examples, call it Blah.jl

> Then ```Oscar.example("Blah.jl")``` will include it, so you don't have to worry about a path
> ```Oscar.revise("Blah.jl")``` will put it under revise, ie. re-load automatically.

Reviewer: ```suggestion
Then `Oscar.example("Blah.jl")` will include it, so you don't have to worry about a path;
and `Oscar.revise("Blah.jl")` will put it under revise, i.e. re-load it automatically whenever you make a change.
```

If Blah.jl starts with

```julia
 module BlahModule
 using Oscar

 ...

 end

 using .BlahModule
``` 

then you can even redefine struct's without quitting Julia (and restarting)

> Julia's convention is to use lower_case_and_underscores, not CamelCase for functions,

Reviewer: ```suggestion
Julia's convention is to use `snake_case`, not `CamelCase` for function names,
```

so that is the accepted style in Oscar.

> types (and their constructor) tend to be CamelCase. However, for me, as a user,

Reviewer: ```suggestion
Types (and their constructor) tend to be in `CamelCase`. However, for me, as a user,
```

please ALSO provide the constructor/ a constructor in underscore_case. As a
user I do usually not know if something is a constructor or a function (nor do I
want to).

In Oscar all objects have a default mathematical meaning and we have multiple dispatch. As
a result
 - IsSurjectiveMapForAbelianGroups
 - IsSurjectiveMapForRing
is not necessary
 - issurjective
will do, the different calling signatures will add the type info. If you have a ring map
and want to test it as a map between additive/ multiplicative groups, then you'll have to 

> create a container for it... (we can always discuss exceptions)

Reviewer: ```suggestion
first convert/coerce/wrap it (in)to a map between groups... (we can always discuss exceptions)
```

> We will have to add "forgetful functor" containers.

Reviewer: ```suggestion
We will have to add "forgetful functor" wrappers.
```
